### PR TITLE
fix abstract nondet exception

### LINKED ISSUE #13945: Dynamic exception type: std::bad_cast
## Description

I followed the [SMTChecker docs](https://docs.soliditylang.org/en/v0.8.18/smtchecker.html#natspec-function-abstraction) and added the following annotation to all of my PRBMath functions:

```solidity
/// @custom:smtchecker abstract-function-nondet
```

See commit [588c91][588c91].

I then proceeded with running the SMTChecker in a private repo that has PRBMath as a dependency, when I got this error:

```text
❯ FOUNDRY_PROFILE=smt forge build --ignored-error-codes 7737
[⠊] Compiling...
[⠊] Compiling 221 files with 0.8.18
[⠒] Solc 0.8.18 finished in 10.34s
Error:
Compiler run failed
Unknown exception during compilation: Dynamic exception type: std::bad_cast
std::exception::what: std::bad_cast
```

Prior to adding the `abstract-function-nondet` annotation, the SMTChecker was running successfully.

## Environment

- Compiler version: 0.8.18
- Target EVM version (as per compiler settings): default
- Framework/IDE (e.g. Truffle or Remix): Forge
- EVM execution environment / backend / blockchain client: REVM
- Operating system: macOS (Apple Silicon)

## Steps to Reproduce

1. Create a Forge project.
2. Install PRBMath from commit [de5ccb9][de5ccb9].
3. Use any PRBMath function.
4. Enable the SMTChecker in the Foundry config e.g. with the config below:

```toml
[profile.smt.model_checker]
  engine = "chc"
  targets = [
    "assert",
    "constantCondition",
    "divByZero",
    "outOfBounds",
    "overflow",
    "underflow",
  ]
```

[588c91]:https://github.com/PaulRBerg/prb-math/commit/588c91

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/13945

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.h [hunk 1]
```diff
@@ -113,9 +113,6 @@ class SMTEncoder: public ASTConstVisitor
 	/// @returns the ModifierDefinition of a ModifierInvocation if possible, or nullptr.
 	static ModifierDefinition const* resolveModifierInvocation(ModifierInvocation const& _invocation, ContractDefinition const* _contract);
 
-	/// @returns the SourceUnit that contains _scopable.
-	static SourceUnit const* sourceUnitContaining(Scopable const& _scopable);
-
```
Comment (ekpyron) [inline libsolidity/formal/SMTEncoder.h:118]: I take it you just happened to notice this is unused and this is in no relation to the fix? Or am I blind at this hour :-)?

**Review by ekpyron — APPROVED** [review]
LGTM

File: libsolidity/formal/SMTEncoder.h (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.h:118]: The bug here was a wrong assumption on `scope()`, so I looked for other uses of `scope()` and found this function, then checked where it's used and it's not so removed it.

**Comment by PaulRBerg** [comment]
Thanks, guys.

Will this fix be included in v0.8.19?

**Comment by leonardoalt** [comment]
@PaulRBerg yep! It'll probably take a few weeks though, considering 0.8.18 was just released. You can ofc use nightlies if you want to test it ahead of the release

**Pull request merged**